### ISSUE TITLE
Further validation of identity and user notices

### ISSUE THREAD

**Issue by smarek** [cakephp/authentication#387]
This is relevant to #316 and #382 

There are multiple issues, that exist with current state of preferred login action, as documented here
https://book.cakephp.org/authentication/2/en/index.html#building-a-login-action

As an example, see this modified/prototype login action
```php
public function login()
{
    // if user already authenticated, redirect them away
    // if user just authenticated eg. request is POST, validate identity
    // further validate identity
    // -> reject the result of authentication process, if the identity fails additional checks
    // -> accept the result of authentication process, if the identity passes all checks
}
```

How would you do this now? example
```php
public function login()
{
  if (!$this->getRequest()->is('post') && $this->Authentication->getIdentity() !== null) {
    $target = $this->Authentication->getLoginRedirect() ?? '/home';
    return $this->redirect($target);
  }

  $result = $this->Authentication->getResult();
  if ($this->getRequest()->is('post')) {
    if (!$result->isValid()) {
      $this->Flash->error('Invalid username or password');
      return;
    }

    $userdata = $result->getData();
    if ($userdata->user_is_enabled !== true) {
      $this->Flash->error('Your account is currently disabled');
      $this->Authentication->logout();
      return;
    }
  }
}
```

The process could be more simple if
  - If the authentication result is valid, the identity was not set unless explicitly accepted by controller
  - If authentication had some method to distinguish between user that is currently logged in and the user that is currently trying to log-in (thus the complicated if check in first line of example `login()`)
  - If the Identifier interface allowed to return validation message as well / if the validation of identified Identity was separate process, that would run only after successful identification process

What do you think?

**Comment by markstory** [comment]
Sounds good. Would you be willing to update the docs with the proposed changes?

**Comment by smarek** [comment]
@markstory sure, i can update current docs with the example, how to further validate and put up warnings/errors/successes

also do you consider updating the library to reflect on the points at the end of my submission?

**Comment by markstory** [comment]
> If authentication had some method to distinguish between user that is currently logged in and the user that is currently trying to log-in (thus the complicated if check in first line of example login())

Each request technically has an authentication attempt as the service will check each authenticator, for example loading data from the session counts as an authentication attempt in the current plugin state. We could make additional affordances for authenticators that scope themselves to a specific URL (like FormAuthenticator). I don't have a good API for that in mind though.

> If the Identifier interface allowed to return validation message as well / if the validation of identified Identity was separate process, that would run only after successful identification process

We currently have error messages available on the `Authenticator\Result` object. I'm not sure what you mea by validation of the identified identity being a separate process though.

**Comment by smarek** [comment]
1. First point was not mentioned, would it be possible to configure Authentication to not automatically set the Identity, unless the Controller accepts the valid Authentication result ?
2. Whole point was to get rid of `(!$this->getRequest()->is('post') && $this->Authentication->getIdentity() !== null)` by using something like `$this->Authentication->identityWasNotCreatedByCurrentRequest()`
3. ↓

  - I see FAILURE_OTHER in ResultInterface, so implementing custom Authenticator (along with custom Identifier) and returning FAILURE_OTHER along with array of errors, that the controller would Flash to frontend might work
  - but for simple validation process, such as checking `UserEntity.is_enabled` this is a lot of work, so see point 1. which could make this easier without need to write as much code
  - **or** if the Authenticators allowed for additional `validators` configuration, that would validate the found identity/entity before returning the `Result` (ie. after check if identity found, before returning Result::SUCCESS, [JWT#L111](https://github.com/cakephp/authentication/blob/master/src/Authenticator/JwtAuthenticator.php#L111), [Form#L129](https://github.com/cakephp/authentication/blob/master/src/Authenticator/FormAuthenticator.php#L129), etc..), that would make the process *separate* and did not force the developer to subclass perfectly working FormAuthenticator or perfectly working PasswordIdentifier or other provided component

**Comment by markstory** [comment]
> First point was not mentioned, would it be possible to configure Authentication to not automatically set the Identity, unless the Controller accepts the valid Authentication result ?

How would the controller 'accept' the authentication result? The authentication middleware finishes running before controllers are even constructed.

> Whole point was to get rid of (!$this->getRequest()->is('post') && $this->Authentication->getIdentity() !== null) by using something like $this->Authentication->identityWasNotCreatedByCurrentRequest()

I don't know if we could have an interface like what you're suggesting as the identity is always *created* on each request. We could potentially include a 'source' on the `Authentication\Result` that would enable checks like

```php
$authResult = $this->request->getAttribute('authenticationResult');
if ($authResult->getSource() === 'Session') {
  // was already authenticated, do a redirect.
}
```

> but for simple validation process, such as checking UserEntity.is_enabled this is a lot of work, so see point 1. which could make this easier without need to write as much code

Additional conditions like this can be done by configuring the `finder` used by the [OrmResolver](https://book.cakephp.org/authentication/2/en/identifiers.html#orm-resolver). You shouldn't need to write a custom identifier and authenticator for this purpose.

**Comment by smarek** [comment]
**Accepting the result**
1) Possibly user callback / subclass of AuthenticationMiddleware, because the validations can imho be done at Middleware level as well
2) It's also probably tight connected to the second point, checking the source of authentication result, controller can now override the identity by setting it through AuthenticationComponent, so `$this->Authentication->setIdentityAccepted(bool)` for controller to suspend/resume the status of successfull authentication could be done

**Verifying source of authentication result**: I very like what you proposed, source would also be useful in more complicated setups for debugging and possibly controller decision, to what intent the identity is trusted

**Validation process**: it cannot be done using finder, because then I'll always end up with FAILURE_IDENTITY_NOT_FOUND
I understand the need to not disclose the existence of Identity in case someone know only eg. email, not the correct password (login form attacks), but once the requester proves he knows valid credentials (email/password) the identity should be found.

Then the validation should occur, if the identity is *acceptable* (disabled, account unpaid, organization suspended, ...), and if validation(s) fail user should be specifically informed about what validation did not succeed, because it's no longer needed to keep the user in-dark about why the authentication does not succeed.

Or is my use-case too alien ? I mean, if you disable user, do you want him to see 'username or password wrong' error message ?

**Comment by markstory** [comment]
> Or is my use-case too alien ? I mean, if you disable user, do you want him to see 'username or password wrong' error message ?

This sounds like something that would exist outside of authentication as a separate middleware/component/controller check. If you don't want to treat disabled users as 'not found' then and instead jail them to a subset of your application that is something that happens after and separately from a successful authentication.

**Comment by DanDrix8391** [comment]
I am facing the same issue, but instead of disabling the User, I have other scenarios:

- two step auth
  after user/password, it redirects to a page to enter a code received in your mobile/email
- require new password
   after login, it redirects to a page to set a new password
- user email verify
   if the user didn't verify its email when he/she signed up (similar to user disabled)
- user have to complete some questions
   if the user tries to login but stopped at the middle of the questions
   in this case I redirect to the step right on the user stopped answering

right now the only way to do is getting the `$this->Authentication->getResult()`, and when it fits any of these cases above, I have to logout the user then do my logic
but I didn't want to logout the user. I like the way the `AuthComponent` was. 
I can simple run the Identify, check if user/password is valid but It doesn't set the Identifier yet, only after `$this->Auth->set($user)`

**Comment by smarek** [comment]
@DanDrix8391 thank you for explaining other cases as well, some of those, will require more work, because filling / validating additional form (through custom Form impl) or suspending the identity activation before next action passes/is taken by user, is not a trivial callback matter.

However I sense here one path, that still would have to be documented/exampled greatly, but that will probably get recommended, and that is using "authorization", because "authentication" takes care of identity, and it's validity (the identity is valid based on set authentication methods) but since the identity is not "authorized" to other actions, than eg. 2FA/MFA or new-password/questionary actions, it should probably be "partially authorized", just enough for the identity to be able to take additional actions, but not enough, to be fully working in app context.

I think "partial authorization" (some of your cases) and "identity valid, but not yet allowed to login" (my case) are different matters, sorry

**Comment by DanDrix8391** [comment]
the last one I agree, it's more authorization then authentication step.
but an user that requires two step authentication, non-verified email or required a new password (then sign in again). It's part of authentication process.

I just wanna a way to verify the user entered the correct user/password without setting the identifier.
Then I'll do my business logic
If the user is good to go, then I do `$this->Authentication->setIdentity($user);`  and everything works as `AuthComponent`

**Comment by markstory** [comment]
Have you considered building an authenticator class that applies the logic you need and treats those as authentication 'failures'? The AuthenticationResult allows you to set error messages/codes that can be used to indicate that the user did not meet all the required steps for authentication. Your login action can use `$result->getErrors()` to control what the user's next step should be. Using an authentication error also removes the need to logout the user and then log them back in again.
